Marlin/Marlin_main.cpp

```diff
@@ -8869,6 +8869,93 @@ inline void gcode_M400() { stepper.synchronize(); }
 
 #endif // FILAMENT_WIDTH_SENSOR
 
+#if ENABLED(REPORT_M408_JSON)
+  /**
+   * M408: Report JSON-style response to serial
+   */
+  inline void gcode_M408() {
+    // I=idle, P=printing from SD card, S=stopped (i.e. needs a reset), C=running config file (i.e starting up), A=paused, D=pausing, R=resuming from a pause, B=busy (e.g. running a macro),
+    SERIAL_PROTOCOLPGM("{\"status\": \"");
+    if (!planner.blocks_queued())
+      SERIAL_PROTOCOLCHAR('I'); // IDLING
+    else
+    if (IS_SD_PRINTING || print_job_timer.isRunning())
+      SERIAL_PROTOCOLCHAR('P'); // SD PRINTING
+    else
+    if (print_job_timer.isPaused())
+      SERIAL_PROTOCOLCHAR('A'); // PAUSED
+    else
+      SERIAL_PROTOCOLCHAR('B'); // SOMETHING ELSE
+    SERIAL_PROTOCOLPGM("\",\"heaters\": [");
+    #if HAS_TEMP_BED
+      SERIAL_PROTOCOL_F(thermalManager.degBed(), 1);
+    #else
+      SERIAL_PROTOCOL(-1);
+    #endif
+    #if HAS_TEMP_HOTEND
+      HOTEND_LOOP() {
+        SERIAL_PROTOCOLCHAR(',');
+        SERIAL_PROTOCOL_F(thermalManager.degHotend(e), 1);
+      }
+    #endif
+    SERIAL_PROTOCOLPGM("],\"active\": [");
+    #if HAS_TEMP_BED
+      SERIAL_PROTOCOL(thermalManager.degTargetBed());
+    #else
+      SERIAL_PROTOCOL(-1);
+    #endif
+    #if HAS_TEMP_HOTEND
+      HOTEND_LOOP() {
+        SERIAL_PROTOCOLCHAR(',');
+        SERIAL_PROTOCOL(thermalManager.degTargetHotend(e));
+      }
+    #endif
+    //skipped standbay & hstat
+    SERIAL_PROTOCOLPGM("],\"pos\": [");
```

Comment:
These custom print statements that create key/value strings should probably be implemented as a small function taking the key and value. The code will likely end up smaller.

Reply:
A single key-value output would be…
```cpp
void floatkey(const char *pstr, const float &val, const bool comma=false) {
  SERIAL_CHAR('"'); serialprintPGM(pstr); SERIAL_CHAR('"');
  SERIAL_CHAR(':'); SERIAL_ECHO(val);
  if (comma) SERIAL_ECHO(',');
}
```
And a function for XYZ elements…
```cpp
void xyzkey(const char *pstr, const float val[XYZ], const bool comma=false) {
  SERIAL_CHAR('"'); serialprintPGM(pstr); SERIAL_CHAR('"');
  SERIAL_ECHO(":[");
  SERIAL_ECHO(val[X_AXIS]); SERIAL_CHAR(',');
  SERIAL_ECHO(val[Y_AXIS]); SERIAL_CHAR(',');
  SERIAL_ECHO(val[Z_AXIS]);
  SERIAL_CHAR(']');
  if (comma) SERIAL_ECHO(',');
}